### PR TITLE
issue 1062: Keep track of (context cancelled) error on connection, and make rows.Next return it

### PR DESCRIPTION
**Pull request by mjl-**
For https://github.com/lib/pq/issues/1062

Instead of just whether the connection is ErrBadConn. Often times, the error
will still be ErrBadConn. But for expired/cancelled contexts, it will be the
error for the context. Most functions still return ErrBadConn per the
database/sql/driver contract ("ErrBadConn should only be returned from [...] a
query method"). For rows.Next() we return the context-related error.

The database/sql/driver contract doesn't look very precise. Is Next a "query
method" and should database/sql handle ErrBadConns when Next returns them?

Do we have more functions that should return the canceled-context error
message?

### CODE REVIEW

File: conn.go [hunk 1]
```diff
@@ -166,6 +167,40 @@ type conn struct {
 	gss GSS
 }
 
+type syncErr struct {
+	err error
```
Comment (otan) [inline conn.go:171]: any reason not to use `type syncErr atomic.Value` here, avoiding the mutex in the impls below?

File: conn_test.go [hunk 2]
```diff
@@ -696,23 +695,19 @@ func TestErrorDuringStartupClosesConn(t *testing.T) {
...
+	cn := conn{}
 	func() {
 		defer cn.errRecover(&err)
 		panic(io.EOF)
 	}()
 	if err != driver.ErrBadConn {
 		t.Fatalf("expected driver.ErrBadConn, got: %#v", err)
 	}
-	if !cn.getBad() {
+	if err := cn.err.get(); err == nil {
```
Comment (otan) [inline conn_test.go:706]: we should assert what kind of error gets returned here

File: conn_test.go [hunk 3]
```diff
@@ -721,7 +716,7 @@ func TestBadConn(t *testing.T) {
...
```
Comment (otan) [inline conn_test.go:719]: here too

File: copy.go [hunk 4]
```diff
@@ -49,12 +49,12 @@ type copyin struct {
 	buffer  []byte
 	rowData chan []byte
 	done    chan bool
-	driver.Result
 
 	closed bool
 
-	sync.Mutex // guards err
+	sync.Mutex // guards err and Result
```
Comment (otan) [inline copy.go:55]: whilst you're here, can you make this easier to not make a mistake by going:
```
type copyin struct {
  // ....
  type mu struct {
     sync.Mutex
     driver.Result
     err error
  }
}

```

then use

```
var c copyin
c.mu.Lock() ....
c.mu.Result() ...
```

that way it's clear how the mutex is used.

File: issues_test.go [hunk 5]
```diff
@@ -58,3 +59,22 @@ func TestIssue1046(t *testing.T) {
...
+	// Ensure that cancelling a QueryRowContext does not result in an ErrBadConn.
+
+	for i := 0; i < 100; i++ {
+		ctx, cancel := context.WithCancel(context.Background())
+		go cancel()
+		row := db.QueryRowContext(ctx, "select 1")
+
+		var v int
+		err := row.Scan(&v)
+		if err == driver.ErrBadConn {
```
Comment (otan) [inline issues_test.go:76]: can we assert instead what error we expect?

**Review by otan — COMMENTED** [review]
i don't check this repo often, sorry for the super slow reply (you'll need to ping me if its >24 business hours)

have a couple of cleanup suggestions for you

**Comment by otan** [comment]
looks like this is also failing tests

File: conn.go (hunk 1 above)
Comment (mjl-) [reply conn.go:171]: IIRC, I tried and ran into not being able to store a nil value. I currently don't see the code storing nil though, so I could have run into it while testing only. I'll give it a try.

**Comment by mjl-** [comment]
The failed error was due to existing checks around "context canceled" not handling the error.
Those places still check for driver.ErrBadConn. I'll look into if I can remove those errors there.

I'll also look into your other cleanup suggestions. Hopefully within the next few days.

File: conn.go (hunk 1 above)
Comment (mjl-) [reply conn.go:171]: ah, the actual reason is that syncErr.set only wants to set an error if it is currently nil. atomic.Value's CompareAndSwap was introduced only in go1.17.

File: copy.go (hunk 4 above)
Comment (mjl-) [reply copy.go]: i added the mu struct. by making the err & Result available under a different name, and fixing the uses, i noticed the usage wasn't correct.

File: conn_test.go (hunk 2 above)
Comment (mjl-) [reply conn_test.go]: fixed

File: conn_test.go (hunk 3 above)
Comment (mjl-) [reply conn_test.go]: also fixed

**Pull request merged**